course7/course_en.md

```diff
@@ -0,0 +1,183 @@
+
+```moonbit
+struct Ref[T] { mut val : T }
+
+fn init {
+  let ref: Ref[Int] = { val: 1 } // ref itself is just a data binding
+  ref.val = 10 // We can modify the fields of the struct
+  println(ref.val.to_string()) // Output 10
+}
+```
+
+Multiple identifiers pointing to the same mutable data structure can be considered aliases, which need to be handled carefully.
+
+```moonbit
+fn alter(a: Ref[Int], b: Ref[Int]) {
+  a.val = 10
+  b.val = 20
+}
+
+fn init {
+  let x: Ref[Int] = { val : 1 }
+  alter(x, x)
+  println(x.val.to_string()) // x.val will be changed twice
+}
+```
+
+## Debugger
+
+Debugging is like being a detective. You're trying to figure out why your program isn't working as expected. MoonBit's debugger is a tool that helps you do this by showing you what's happening inside your program as it runs. You can pause the program at any point, look at the values of variables, and step through the code one line at a time. This is incredibly useful for understanding complex behavior and fixing bugs.
+
+![](../pics/debugger.png)
+
+## Loops
+
+Loops are a way to repeat a block of code multiple times. In MoonBit, we can use a `while` loop to do this. We start by defining a loop variable, say `i`, and giving it an initial value. Then we check a condition—if the condition is true, we execute the commands inside the loop, and then we update the loop variable. This process repeats until the condition is no longer true, at which point the loop ends.
+
+```moonbit
+let mut i = 0
+while i < 2 {
+  println("Output")
+  i += 1
+} // Repeat output 2 times
+```
+
+For example, we might want to print the numbers from `0` to `1`. We can do this with a loop that checks if `i` is less than `2`. If it is, we print the current value of `i`, increment `i` by `1`, and then repeat the process. The loop continues until `i` is no longer less than `2`, and then we're done.
+
+## Loops and Recursion
+
+Loops and recursion are equivalent. A loop is a set of instructions that we repeat until a certain condition is met, while recursion is a way of solving a problem by breaking it down into smaller, similar problems. In programming, we can often rewrite a loop as a recursive function and vice versa.
+
+For example, let's consider calculating the Fibonacci sequence. We can do this with a loop that keeps track of the last two numbers and updates them as it goes. Alternatively, we can write a recursive function that calls itself with smaller numbers until it reaches the base cases (0 and 1).
+
+```moonbit
+// Loop form
+fn fib_mut(n: Int) -> Int {
+  let mut a = 0;
+  let mut b = 1;
+  let mut i = 0
+  while i < n {
+    let t = a + b
+    a = b; b = t
+    i += 1
+  }
+  a
+}
+
+// Recursive form
```

Comment:
Not equivalent. If you don't know how to write an equivalent version, just use the same example as the presentation.

Reply:
the output of these two functinos is equivalent

Reply:
The purpose of this section is to describe how to write the loop with a tail recursion function.

Reply:
> The purpose of this section is to describe how to write the loop with a tail recursion function.

but why put a Fibonacci function as example here inside chinese version

Reply:
I didn't put a Fibonacci function here, that's also why I said you may use the example as the presentation. Also, since you've mentioned, you forgot to put the fibonacci example in the 'Mutable data' section (which was to demonstrate that having mutable data structure can also lead to referential transparency)